File: src/main/kotlin/com/miumiuhaskeer/fastmessage/statistic/model/UserDetailsImpl.java
package com.miumiuhaskeer.fastmessage.statistic.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.miumiuhaskeer.fastmessage.statistic.model.entity.User;
import lombok.Data;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Class copied from FastMessage project
 */
@Data
public class UserDetailsImpl extends ExtendedUserDetails {

    private final Long id;
    private final String email;

    @JsonIgnore
    private final String password;

    private final Collection<? extends GrantedAuthority> authorities;

    public static UserDetailsImpl from(User user) {
        List<GrantedAuthority> authorities = user.getRoles().stream()
                .map(role -> new SimpleGrantedAuthority(role.getName().name()))
                .collect(Collectors.toList());

        return new UserDetailsImpl(
                user.getId(),
                user.getEmail(),
                user.getPassword(),
                authorities
        );
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/kotlin/com/miumiuhaskeer/fastmessage/statistic/model/UserDetailsImpl.kt
package com.miumiuhaskeer.fastmessage.statistic.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.miumiuhaskeer.fastmessage.statistic.model.entity.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import java.util.stream.Collectors

class UserDetailsImpl(
    override val id: Long?,
    override val email: String?,
    @JsonIgnore private val password: String?,
    private val authorities: Collection<GrantedAuthority?>
): ExtendedUserDetails() {

    companion object {

        // TODO remove after migrating to Kotlin
        @JvmStatic
        fun from(user: User): UserDetailsImpl {
            val authorities: List<GrantedAuthority> = user.roles.stream().map {
                    SimpleGrantedAuthority(it.name!!.name)
                }.collect(Collectors.toList())

            return UserDetailsImpl(
                user.id,
                user.email,
                user.password,
                authorities
            )
        }
    }

    override fun getPassword() = password
    override fun getAuthorities() = authorities

    override fun isAccountNonExpired() = true
    override fun isAccountNonLocked() = true
    override fun isCredentialsNonExpired() = true
    override fun isEnabled() = true
}